certvfy/sigvfycert.c

```diff
@@ -0,0 +1,230 @@
+ * (at your option) any later version.
+ *
+ * wolfSSL is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#include <stdio.h>
+
+#ifdef HAVE_CONFIG_H
+    #include <config.h>
+#endif
+
+#ifndef WOLFSSL_USER_SETTINGS
+    #include <wolfssl/options.h>
+#endif
+#include <wolfssl/wolfcrypt/settings.h>
+
+#include <wolfssl/wolfcrypt/asn_public.h>
+#include <wolfssl/wolfcrypt/asn.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/rsa.h>
+#include <wolfssl/wolfcrypt/signature.h>
+
+#define MAX_DER_SZ          4096
+
+int load_file(const char* name, byte* buf, int bufSz)
+{
+    FILE* file;
+
+    file = fopen(name, "rb");
+    if (file == NULL) {
+        return 0;
+    }
+    bufSz = fread(buf, 1, bufSz, file);
+    fclose(file);
+
+    return bufSz;
+}
+
+int get_rsa_public_key_from_ca(const char* caCert, RsaKey* rsaKey)
+{
+    int res = 0;
+    int ret;
+    byte caDer[MAX_DER_SZ];
+    int caDerSz;
+    DecodedCert ca;
+    word32 idx;
+
+    XMEMSET(&ca, 0, sizeof(ca));
+
+    /* Load the DER encoded CA certificate. */
+    caDerSz = load_file(caCert, caDer, (int)sizeof(caDer));
+    if (caDerSz == 0) {
+        printf("Failed to load CA file\n");
+        res = 1;
+        goto exit;
+    }
+
+    /* Put the CA certificate data into the object. */
+    wc_InitDecodedCert(&ca, caDer, caDerSz, NULL);
+    /* Parse fields of the certificate. */
+    ret = wc_ParseCert(&ca, CERT_TYPE, 0, NULL);
+    if (ret != 0) {
+        printf("Parsing CA failed: %s (%d)\n", wc_GetErrorString(ret), ret);
+        res = 1;
+        goto exit;
+    }
+
+    /* Initialize an RSA key object. */
+    ret = wc_InitRsaKey(rsaKey, NULL);
+    if (ret != 0) {
+        printf("Initializing RSA key failed: %s (%d)\n", wc_GetErrorString(ret),
+            ret);
+        res = 1;
+        goto exit;
+    }
+
+    /* Decode the DER into RSA key object. */
+    idx = 0;
+    ret = wc_RsaPublicKeyDecode(ca.publicKey, &idx, rsaKey, ca.pubKeySize);
+    if (ret != 0) {
+        printf("Decoding RSA key failed: %s (%d)\n", wc_GetErrorString(ret),
+            ret);
+        res = 1;
+        goto exit;
+    }
+
+exit:
+    wc_FreeDecodedCert(&ca);
+
+    return res;
+}
+
+int get_cert_tbs_sig(byte* certDer, int certDerSz, byte** tbs, int* tbsSz,
+    byte** sig, int* sigSz)
+{
+    int res = 0;
+    int ret;
+    word32 tbsIdx = 0;
+    word32 idx;
+    word32 sigIdx;
+    int tbsLen;
+    int sigAlgLen;
+    int sigLen;
+    int len;
+
+    /* Parse outer SEQUENCE. */
+    if ((ret = GetSequence(certDer, &tbsIdx, &len, certDerSz)) < 0) {
```

Comment:
`GetSequence` is not a public API. Is there another way to accomplish this?

Reply:
Added implementations of functions not available.